File: Online Judges/Huxley/595DuasExtremidades.py
inf = 2**33
DEBUG = 1

def solve(dp, cards, lo, hi):
    #print(lo, hi)
    if (lo >= hi):
        return(0)

    if (dp[lo][hi] == -1):
        aux = -inf
        me = cards[lo]
        #if (lo + 1 < len(cards)):
        if (cards[lo + 1] >= cards[hi]):
            greedy = cards[lo + 1]
            aux = max(aux, solve(dp, cards, lo + 2, hi) + (me - greedy))
        else:
            greedy = cards[hi]
            aux = max(aux, solve(dp, cards, lo + 1, hi - 1) + (me - greedy))
        me = cards[hi]
        #if (hi - 1 >= 0):
        if (cards[lo] >= cards[hi - 1]):
            greedy = cards[lo]
            aux = max(aux, solve(dp, cards, lo + 1, hi - 1) + (me - greedy))
        else:
            greedy = cards[hi - 1]
            aux = max(aux, solve(dp, cards, lo, hi - 2) + (me - greedy))
        dp[lo][hi] = aux

    return(dp[lo][hi])

run = 0
while (True):
    run += 1
    line = list(map(int, input().split()))
    qCards, cards = line[0], line[1:]
    if (qCards == 0):
        break
    dp = [[-1] * qCards for i in range(qCards)]
    answer = solve(dp, cards, 0, qCards - 1)
    #print(dp)
    print("In game ", run, ", the greedy strategy might lose by as many as ", answer, " points.", sep='')

File: Online Judges/Huxley/595DuasExtremidades.cpp
#include <bits/stdc++.h>
#define lli long long int
using namespace std;

const int maxN = 1e3; int n;
const lli inf = 1e9;
int a[maxN];

lli memo[maxN][maxN][2];
lli solve(int lo = 0, int hi = n - 1, bool turn = true)
{
  if (lo > hi) return 0;
  if (memo[lo][hi][turn] == -1)
  {
    lli ans;
    if (turn)
      ans = max(solve(lo + 1, hi, !turn) + a[lo], solve(lo, hi - 1, !turn) + a[hi]);
    else
      ans = a[lo] >= a[hi] ? solve(lo + 1, hi, !turn) - a[lo] : solve(lo, hi - 1, !turn) - a[hi];
    memo[lo][hi][turn] = ans;
  }
  return memo[lo][hi][turn];
}

int main()
{
  int t = 1;
  while (scanf("%d", &n) && n)
  {
    // printf("%d\n", n); fflush(stdout);
    for (int i = 0; i < n; i ++)
      scanf("%d", &a[i]);

    for (int i = 0; i < n; i ++) for (int j = 0; j < n; j ++) memo[i][j][0] = memo[i][j][1] = -1;
    // memset(memo, -1, sizeof(memo));
    lli ans = solve();
    printf("In game %d, the greedy strategy might lose by as many as %lld points.\n", t ++, ans);
  }
  return(0);
}